File: frontend/src/features/masters/models/customer.ts
import type { PaginatedResponse } from './common';

export type CustomerStatus = 'Active' | 'Inactive';

export interface Customer {
  id: string;
  customer_code: string;
  customer_name: string;
  address: string | null;
  gstn_number: string | null;
  contact_person: string | null;
  contact_number: string | null;
  contact_email: string | null;
  status: CustomerStatus;
  created_at: string;
  modified_at: string;
  created_by: string | null;
  modified_by: string | null;
}

export interface CustomerListResponse extends PaginatedResponse<Customer> {}

export interface CreateCustomerRequest {
  customer_code: string;
  customer_name: string;
  address?: string;
  gstn_number?: string;
  contact_person?: string;
  contact_number?: string;
  contact_email?: string;
}

export interface UpdateCustomerRequest extends Partial<CreateCustomerRequest> {
  status?: CustomerStatus;
}

File: backend/src/domain/entities/masters/customer.py
"""
Customer (serviced/invoiced institution) domain entity.

Represents a government body, hospital, pharmacy, or institution serviced and
invoiced by a vendor. Inherits identity and audit fields from ``BaseEntity``.
"""

from dataclasses import dataclass, field

from src.domain.entities.base_entity import BaseEntity
from src.domain.enums.masters import CustomerStatus


@dataclass
class CustomerEntity(BaseEntity):
    """
    Customer Master aggregate root.

    Business Rules:
    - Customer Code and Customer Name are required (validated at service level).
    - Customer Code must be unique (enforced at repository/DB level).
    - GSTN Number, when supplied, must match the pattern ``[A-Z0-9]{15}``.
    - New customers default to ``Active`` status.
    - A Customer cannot be deleted while referenced by an active
      Vendor-Customer Mapping or any Invoice Header (enforced at service level
      using repository reference-existence checks).
    """

    customer_code: str = field(default="")
    customer_name: str = field(default="")
    address: str | None = field(default=None)
    gstn_number: str | None = field(default=None)
    contact_person: str | None = field(default=None)
    contact_number: str | None = field(default=None)
    contact_email: str | None = field(default=None)
    status: CustomerStatus = field(default=CustomerStatus.Active)
